### ISSUE TITLE
Generator[None, Any, None] inferred incorrectly

### ISSUE THREAD

**Issue by CGamesPlay** [microsoft/pyright#8127]
**Describe the bug**
PyRight is able to detect that all `yield` statements in a generator are dead code, however it infers the yield type of such a generator to be `None`, when it should infer it to be `Never`. This is relevant for higher-order functions which accept such generators as parameters.

Note: the behavior is identical for `AsyncGenerator` as well (it infers `AsyncGenerator[None, Any]` when it should infer `AsyncGenerator[Never, Any]`).

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSAIYy4A0UAwuQDaPkBGjRtAckQG5m0QfIkwD68BEQBQUgCZFgUAK4BnIqLQkAFJpQAuQiTKVcAbVQxuYUldIBdAJR6pUV1GC4oKJdFRRdzm5BUAggFlreEA4ycgr%2BJACMWk4ubnBIRIyyUAky8oq6AEzJgW5CMEogKKmu6ZnZuVK6AMwGDMxsHKamdrTEpBRUIKY8-CC2nFBc1kR2dlAAvPEohVAAxLCIRAZIaCi40lJC-GISRDqJ0ccijOJbFytXwqf3LdFSquq6D0kO0Z8abRFZL-NSAlAPZogoA)

```python
from typing import Generator, Callable, Never, reveal_type

def use_gen(gen: Generator[int, None, None]):
    for num in gen:
        print(num)


def gen1():
    yield 1

def gen2():
    return
    yield 1

gen3: Callable[[], Generator[Never, None, None]] = gen2 # type: ignore

reveal_type(gen1) # Type of "gen1" is "() -> Generator[Literal[1], Any, None]"
reveal_type(gen2) # Type of "gen2" is "() -> Generator[None, Any, None]"
reveal_type(gen3) # Type of "gen3" is "() -> Generator[Never, None, None]"

use_gen(gen1())
use_gen(gen2()) # Error
use_gen(gen3())
```

```
Argument of type "Generator[None, Any, None]" cannot be assigned to parameter "gen" of type "Generator[int, None, None]" in function "use_gen"
  "Generator[None, Any, None]" is incompatible with "Generator[int, None, None]"
    Type parameter "_YieldT_co@Generator" is covariant, but "None" is not a subtype of "int"
      "None" is incompatible with "int"  (reportArgumentType)
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version? See playground.

**Comment by erictraut** [comment]
It's not obvious to me that the yield type should be inferred as `Never` when all of the yield statements are unreachable, but I think you've made a reasonable case. For comparison, if there are no reachable `return` statements, the inferred return type is `None` unless it can be determined that the end of the function is unreachable, in which case the inferred return type is `Never`. 

I'll make your proposed change and see if it breaks assumptions in existing typed code bases.

**Comment by erictraut** [comment]
It looks like this had no impact on existing typed code bases, so I'm merging the proposed change. This will be included in the next pyright release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.368.
